### PR TITLE
Handle new connections during shutdown

### PR DESCRIPTION
**Pull request by alfonsoros88**
Currently it is possible to accept new connections during a shutdown.
This causes a sort of dead lock where the shutdown cannot terminate as
log as connections remain open but requests cannot be answer because the
runtime is just waiting for the shutdown.

To handle this situation, if a shutdown request is issued, the console
is closed for new connections and request in the event loop are
ignored.

### CODE REVIEW

File: northstar/src/runtime/console.rs [hunk 1]
```diff
@@ -357,15 +357,16 @@ impl Console {
...
+                    trace!("{:?} -> event loop", request);
+                    let response = match event_tx.send(Event::Console(request, reply_tx)).then(|_| reply_rx).await {
+                        Ok(response) => response,
+                        Err(e) => {
+                            // This could happen if the runtime is in process of shutting down
+                            warn!("Failed to get response from event loop: {}", e);
+                            break;
+                        }
+                    };
+                    trace!("{:?} <- event loop", response);
```
Comment (flxo) [inline northstar/src/runtime/console.rs:369]: If the connections are closed, before everything else it shut down the `expects` are fine. This patch logs things but don't fail on the internal bug when the main loop is down or the runtime core doesn't respond.

File: northstar/src/runtime/minijail.rs [hunk 2]
```diff
@@ -69,16 +70,25 @@ impl<'a> Minijail<'a> {
...
+                        match line.chars().next() {
+                            Some('D') => debug!("{}", l),
+                            Some('I') => info!("{}", l),
+                            Some('W') => warn!("{}", l),
+                            Some('E') => error!("{}", l),
+                            _ => trace!("{}", line),
+                        }
+                    }
+                    _ = task_stop.cancelled() => break,
+                    else => break,
```
Comment (flxo) [inline northstar/src/runtime/minijail.rs:91]: A nice pattern is to avoid the need to find a name doesn't fit (just because a clone is needed). Put it into braces. Clone everything needed on the top. This also underlines what elements are moved into the task.
```rust
{
   let stop_token = stop_token.clone();
   ....
}
```

File: northstar/src/runtime/mod.rs [hunk 3]
```diff
@@ -206,7 +206,18 @@ async fn runtime_task(config: &'_ Config, stop: CancellationToken) -> Result<(),
...
             // The runtime os commanded to shut down and exit.
-            Event::Shutdown => break state.shutdown().await,
+            Event::Shutdown => {
+                log::debug!("Stopping to process events");
+                drop(event_rx);
+
+                log::debug!("Stopping console");
+                if let Some(console) = console.take() {
+                    console.shutdown().await.map_err(Error::Console)?;
+                }
```
Comment (flxo) [inline northstar/src/runtime/mod.rs:216]: Shouldn't this be the other way round. Stop the console and it's connections *before* the main loop?

File: northstar/src/runtime/process.rs [hunk 4]
```diff
@@ -93,9 +93,13 @@ pub(crate) async fn waitpid(
...
-        event_handle
-            .blocking_send(Event::Exit(container, status.clone()))
-            .expect("Internal channel error on main event handle");
+        //
+        // TODO this is needed so that the runtime does not panic on shutdown
+        //
+        match event_handle.blocking_send(Event::Exit(container, status.clone())) {
+            Ok(()) => (),
+            Err(e) => log::warn!("Failed to send container exit status: {}", e),
+        };
```
Comment (flxo) [inline northstar/src/runtime/process.rs:102]: If this is just logged there's a optional hard error that leaks a process ignored. This error must panic!

File: northstar/src/runtime/state.rs [hunk 5]
```diff
@@ -756,10 +757,8 @@ impl<'a> State<'a> {
     }
 
     async fn notification(&self, n: Notification) {
-        self.events_tx
-            .send(Event::Notification(n))
-            .await
-            .expect("Internal channel error on main");
+        // TODO do not ignore this
+        let _ = self.events_tx.send(Event::Notification(n)).await;
```
Comment (flxo) [inline northstar/src/runtime/state.rs:761]: Sending on `event_tx` must be an error.

File: northstar/src/runtime/mod.rs (hunk 3 above)
Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:216]: The event loop has to be stopped first. This _signals_ the open connections that they won't be getting their responses. Then the next step is to wait for those connections to terminate and close the console. Finally the state is shut down.

File: northstar/src/runtime/minijail.rs (hunk 2 above)
Comment (alfonsoros88) [reply northstar/src/runtime/minijail.rs:91]: Sure, but this function has already too many levels :sweat_smile:

File: northstar/src/runtime/console.rs (hunk 1 above)
Comment (alfonsoros88) [reply northstar/src/runtime/console.rs:369]: The problem are the connections that are at this point waiting for a response. They won't close and therefore won't permit that the console is finally closed. That's the actual issue here.

File: northstar/src/runtime/process.rs (hunk 4 above)
Comment (alfonsoros88) [reply northstar/src/runtime/process.rs:102]: Yes, I need to fine a solution for this one.

File: northstar/src/runtime/mod.rs (hunk 3 above)
Comment (marcmo) [reply northstar/src/runtime/mod.rs:216]: Conceptually the event loop should remain processing until no further events are generated. 
We need another way to signal open connections that will not be serviced anymore.

File: northstar/src/runtime/console.rs (hunk 1 above)
Comment (flxo) [reply northstar/src/runtime/console.rs:369]: Is the "waiting for a response" uninterruptible? Should be doable with the `CancelationToken` you pass into the connection loop.

File: northstar/src/runtime/minijail.rs (hunk 2 above)
Comment (flxo) [reply northstar/src/runtime/minijail.rs:91]: But short ones ;-)

File: northstar/src/runtime/mod.rs (hunk 3 above)
Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:216]: There is this `shutdown` method on the connections, I will try to invert the order with that.

Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:216]: @marcmo, what should be the behavior with a shutdown request followed by n other requests? the remaining request should be processed before the shutdown?

File: northstar/src/runtime/minijail.rs (hunk 2 above)
Comment (alfonsoros88) [reply northstar/src/runtime/minijail.rs:91]: as you wish :)

**Comment by flxo** [comment]
@alfonsoros88 I think the shutdown procedure can be simplified by closing any open connections and listeners (and wait for them as you do) *before* stopping the state. I sketched this in https://github.com/flxo/northstar/tree/connections_during_shutdown
There can't be any console commands in flight when the `Shutdown` event is processes so from my understanding closing the console must be enough to prevent the initial problem that this PR tries to solve.

**Comment by alfonsoros88** [comment]
@flxo I had a look to your suggestion but it seems that after receiving a shutdown request it would stop processing events right after that. @marcmo suggested that we should handle those remaining events in the queue like exits and notifications before closing it. Also when the state shuts down, it potentially produces more of those events.

File: northstar/src/runtime/minijail.rs [hunk 6]
```diff
@@ -469,8 +470,12 @@ impl Process {
     /// it is SIGKILLed.
     pub async fn terminate(&mut self, timeout: time::Duration) -> Result<ExitStatus, Error> {
         debug!("Sending SIGTERM to {}", self.pid);
-        signal::kill(unistd::Pid::from_raw(self.pid as i32), Some(SIGTERM))
-            .map_err(|e| Error::Os(format!("Failed to SIGTERM {}", self.pid), e))?;
+
+        match signal::kill(unistd::Pid::from_raw(self.pid as i32), Some(SIGTERM)) {
+            Err(nix::Error::Sys(Errno::ESRCH)) => return (&mut self.exit_status).await,
```
Comment (flxo) [inline northstar/src/runtime/minijail.rs:475]: `ESRCH` means that libc couldn't find a matching process. Isn't this an error condition? We're trying to terminate a process and it's not found. Must be something wrong in the state.

**Comment by flxo** [comment]
> @flxo I had a look to your suggestion but it seems that after receiving a shutdown request it would stop processing events right after that. @marcmo suggested that we should handle those remaining events in the queue like exits and notifications before closing it. Also when the state shuts down, it potentially produces more of those events.

The events that might be in the queue don't make sense when the runtime already know that it's gonna shut down. E.g start soemthing. I think also the shutdown notification is nonsense since this event is signaled well via the graceful shutdown of the connection.

File: northstar/src/runtime/minijail.rs (hunk 6 above)
Comment (alfonsoros88) [reply northstar/src/runtime/minijail.rs:475]: Peter found this issue, this is a race if where the process can exit just before this point in which case its pid is no longer valid. Maybe I should put this as a comment.

**Comment by alfonsoros88** [comment]
@flxo I wouldn't say that it is nonesensical, we handle such events in `State` with the intention to cleanup after containers. Currently we are not leaking things like cgroups because for example both `stop` and `on_exit` call `destroy()` but probably the correct would be to let the cleanup to `on_exit`. In that case we probably would like to handle the remaining events. 

It is similar for the notifications, there is some sense to it. From the point of view of the client there is no difference between a graceful shutdown and a broken connection. The client would probably be interested in listening notifications related to those containers it started.

**Comment by flxo** [comment]
> @flxo I wouldn't say that it is nonesensical, we handle such events in `State` with the intention to cleanup after containers. Currently we are not leaking things like cgroups because for example both `stop` and `on_exit` call `destroy()` but probably the correct would be to let the cleanup to `on_exit`. In that case we probably would like to handle the remaining events.

If the "cleanup" would be just done on `on_exit` the shutdown would need to be synchronised to the last `on_exit` event etc. The topic here's asynchronous stop: Stop containers in parallel like the start. We do not do that currently because the miliseconds don't count on the stopping case as they do when starting.

> It is similar for the notifications, there is some sense to it. From the point of view of the client there is no difference between a graceful shutdown and a broken connection.

There is a difference.

 > The client would probably be interested in listening notifications related to those containers it started.

I don't get this. Why should a client want to wait for notifications after sending a shutdown?

File: northstar/src/runtime/minijail.rs (hunk 6 above)
Comment (flxo) [reply northstar/src/runtime/minijail.rs]: Ok! Thank for pointing out.

A comment would be great.

**Comment by alfonsoros88** [comment]
> If the "cleanup" would be just done on `on_exit` the shutdown would need to be synchronised to the last `on_exit` event etc. The topic here's asynchronous stop: Stop containers in parallel like the start. We do not do that currently because the miliseconds don't count on the stopping case as they do when starting.

Probably  I'm misunderstanding something but I don't see what's the problem. Still sounds like handling the "cleanup" in `on_exit` is the right thing to do. And if as a side product that allows us to have a parallel stopping of the containers then why not? :sweat_smile: 

> There is a difference.

If I fire 2 instances of `nstar` for example, and call `shutdown` from one of them, the message I receive in both instances is "Disconnected, trying to reconnect". Perhaps we are referring to different things.

> I don't get this. Why should a client want to wait for notifications after sending a shutdown?

I meant that if at some point, a user is scripting some client using `northstar::api::client`, I would probably be interested in the notifications, even those sent during shutdown. Again, maybe because currently there seems to be no way to distinguish between a lost of connectivity or a shutdown procedure.

File: northstar/src/runtime/minijail.rs [hunk 7]
```diff
@@ -82,19 +83,30 @@ impl Launcher for Minijail {
...
+                                Some('D') => debug!("{}", l),
+                                Some('I') => info!("{}", l),
+                                Some('W') => warn!("{}", l),
+                                Some('E') => error!("{}", l),
+                                _ => trace!("{}", line),
+                            }
+                        }
+                        _ = stop_token.cancelled() => break,
+                        else => break,
+                    }
```
Comment (flxo) [inline northstar/src/runtime/minijail.rs:106]: Think this is unrelated to the PR.

File: northstar/src/runtime/mod.rs [hunk 8]
```diff
@@ -272,42 +272,68 @@ async fn runtime_task(config: &'_ Config, stop: CancellationToken) -> Result<(),
...
+                // debug!("Stop accepting new connections");
+                // if let Some(console) = &console {
+                //     console.stop_new_connctions();
+                // }
+
+                debug!("Shutting down runtime");
+                state.shutdown().await.unwrap();
+
+                debug!("Closing event loop");
+                event_rx.close();
```
Comment (flxo) [inline northstar/src/runtime/mod.rs:315]: Closing `event_rx` will make this loop break. The flag `shutdown_issued` is never read with value `true`.

File: northstar_tests/tests/tests.rs [hunk 9]
```diff
@@ -347,6 +348,50 @@ test!(connect_version, {
...
+                    .ok(),
+                Err(_) => break,
+            };
+        }
+    });
+
+    let result = runtime.shutdown().await;
+
+    let (fst, snd) = tokio::join!(start_container, stop_container);
+    assert!(fst.is_ok() && snd.is_ok());
```
Comment (flxo) [inline northstar_tests/tests/tests.rs:390]: This just checks for spawn errors. The closures in the task return `()`.

**Review by flxo — COMMENTED** [review]
I like the idea to refactor the listeners into an enum - see comments on the `accept`.

The test is broken and the `shutdown_issued` flag seems wrong to me. Maybe I didn't get the intention.

I'd really vote for closing the listeners and connection upon a shutdown and that's it. That is a graceful close which can be identified by clients.

File: northstar/src/runtime/console.rs [hunk 10]
```diff
@@ -93,130 +101,18 @@ impl Console {
...
-                self.tasks.push(task);
-            }
-            _ => unreachable!(),
-        }
+        let task = match Listener::new(&self.url)
+            .await
+            .map_err(|e| Error::Io("Failed to remove unix socket".into(), e))?
+        {
+            Listener::Tcp(listener) => task::spawn(async move {
+                handle_connections(|| listener.accept(), event_tx, notification_tx, stop).await
```
[the remark this answers is not in the document]
Comment (alfonsoros88) [reply northstar/src/runtime/console.rs:109]: yes, `handle_connections` is has a loop to handle new connections.

File: northstar/src/runtime/console.rs [hunk 11]
```diff
@@ -69,6 +69,14 @@ pub enum Error {
     Protocol(String),
     #[error("IO error: {0} ({1})")]
     Io(String, #[source] io::Error),
+    #[error("Event loop closed, request cannot be processed")]
+    EventLoopClosed,
+}
+
+macro_rules! io_err {
```
[the remark this answers is not in the document]
Comment (alfonsoros88) [reply northstar/src/runtime/console.rs:76]: :+1:  didn't know that

File: northstar/src/runtime/minijail.rs (hunk 7 above)
Comment (alfonsoros88) [reply northstar/src/runtime/minijail.rs:106]: yes, partially, I needed this at some point to be able to stop the `tokio` runtime after the shutdown.

File: northstar/src/runtime/console.rs [hunk 12]
```diff
@@ -452,6 +268,197 @@ impl Console {
...
+/// Generic handling of connections. The first parameter is a function that when called awaits for
+/// a new connection. The connections are represented as a pair of a stream and some client
+/// identifier.
+///
+/// All the connections container stored the tasks corresponding to each active connection. As
+/// these tasks terminate, they are removed from the connections container. Once a stop is issued,
+/// the termination of the remaining connections will be awaited.
+///
+async fn handle_connections<AcceptConnection, Connection, Stream, Client, E>(
+    accept: AcceptConnection,
```
[the remark this answers is not in the document]
Comment (alfonsoros88) [reply northstar/src/runtime/console.rs:397]: I'm not sure if it has to be a stream, it is _something_ that awaits for a new connection.

File: northstar/src/runtime/mod.rs (hunk 8 above)
Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:315]: https://docs.rs/tokio/1.5.0/tokio/sync/mpsc/struct.Receiver.html#method.close The documentation mentions that a call to `close` will not drop the receiver, i.e. all the events generated in `state.shutdown()` will remain there (e.g. all the stop containers events and notifications). The `shutdown_issued` flag is used in the first arm of this match expression to reply to all those possibly queued console requests with an error.

File: northstar_tests/tests/tests.rs (hunk 9 above)
Comment (alfonsoros88) [reply northstar_tests/tests/tests.rs:390]: yes, that's fine. This just checks that those results are ok.

**Comment by alfonsoros88** [comment]
> I like the idea to refactor the listeners into an enum - see comments on the `accept`.
> 
> The test is broken and the `shutdown_issued` flag seems wrong to me. Maybe I didn't get the intention.
> 
> I'd really vote for closing the listeners and connection upon a shutdown and that's it. That is a graceful close which can be identified by clients.

Well yes, at this point this pr is just a scrap yard, the question is if there is still anything of value here. If we want to take the refactorings or not and if we want to change the order of things in the shutdown or not. I will stick to whatever @flxo  and @marcmo  decide.

File: northstar/src/runtime/console.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (flxo) [reply northstar/src/runtime/console.rs:109]: But doesn't a `listener.accept()` actually just return a future because of of 

```rust
pub async fn accept(&self) -> Result<(TcpStream, SocketAddr)>
```
?

My understanding about futures is that it only can resolve once e.g: [play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9f9d15b6b620576b67202799088857ca)

I would get it sorted if the the argument would be `listener.accept` without `()`.

File: northstar/src/runtime/console.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (flxo) [reply northstar/src/runtime/console.rs:76]: Quite new - you would have seen it during a rebase.

We should have done something like this macro or the fn earlier because this `String` sucks.

**Comment by flxo** [comment]
I'd like to merge the Listener implication because it removes slightly duplicate code.  Same for the minijail log shutdown which can't do any harm. 

I have still a hard time to not just close the connections and send "we're closing soon and ignore your request" before closing the connection. There's no win for the client.

File: northstar/src/runtime/console.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (alfonsoros88) [reply northstar/src/runtime/console.rs:109]: The first argument is actually a closure that calls `accept`. The closure is called inside the loop in `handle_connections`. It's kind of a trick that I used to avoid writing a new trait for `UnixListener` and `TcpListener` and just use a `AcceptConnection: Fn() -> NewConnection` generic.

File: northstar/src/runtime/console.rs [hunk 13]
```diff
@@ -93,130 +101,18 @@ impl Console {
...
-        }
+        let task = match Listener::new(&self.url)
+            .await
+            .map_err(|e| Error::Io("Failed to remove unix socket".into(), e))?
+        {
+            Listener::Tcp(listener) => task::spawn(async move {
+                handle_connections(|| listener.accept(), event_tx, notification_tx, stop).await
+            }),
+            Listener::Unix(listener) => task::spawn(async move {
+                handle_connections(|| listener.accept(), event_tx, notification_tx, stop).await
```
Comment (marcmo) [inline northstar/src/runtime/console.rs:112]: quite elegant how you removed the code duplication here

File: northstar/src/runtime/console.rs [hunk 14]
```diff
@@ -452,6 +268,197 @@ impl Console {
...
+    notification_tx: broadcast::Sender<Notification>,
+    stop: CancellationToken,
+) where
+    AcceptConnection: Fn() -> Connection,
+    Connection: Future<Output = Result<(Stream, Client), E>>,
+    Stream: AsyncWrite + AsyncRead + Unpin + Send + 'static,
+    Client: Into<ClientId>,
+    E: std::fmt::Debug,
+{
+    let mut connections = FuturesUnordered::new();
```
Comment (marcmo) [inline northstar/src/runtime/console.rs:408]: didn't know about this `FuturesUnordered`...where did you learn about that?

File: northstar/src/runtime/minijail.rs (hunk 7 above)
Comment (marcmo) [reply northstar/src/runtime/minijail.rs:106]: makes sense

File: northstar/src/runtime/console.rs [hunk 15]
```diff
@@ -452,6 +268,197 @@ impl Console {
...
+            send_event(&event_loop, Request::Message(message)).await?
+        };
+
+    Ok(api::model::Message {
+        id: message_id,
+        payload: api::model::Payload::Response(response),
+    })
+}
+
+/// Sends an `Event::Console` to the event loop and await its response
```
Comment (marcmo) [inline northstar/src/runtime/console.rs:336]: more clear now

**Comment by marcmo** [comment]
looks rather good for my taste...I'd merge it

File: northstar/src/runtime/console.rs (hunk 14 above)
Comment (alfonsoros88) [reply northstar/src/runtime/console.rs:408]: this one is used already in master, @flxo  introduced it in the last update to `console.rs`

**Comment by marcmo** [comment]
oh...just saw that this pr seems to be obsolete and you added #331 instead? if so let's close this @alfonsoros88
